hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDeploymentManagement.java

```diff
@@ -630,41 +639,43 @@ private JpaAction closeActionIfSetWasAlreadyAssigned(final JpaAction action) {
         return action;
     }
 
-    private JpaAction startScheduledActionIfNoCancelationHasToBeHandledFirst(final JpaAction action) {
-        // check if we need to override running update actions
-        final List<Long> overrideObsoleteUpdateActions;
+    private List<Action> startScheduledActionsAndHandleOpenCancellationFirst(final List<JpaAction> actions) {
+        if (!isMultiAssignmentsEnabled()) {
+            closeOrCancelOpenDeviceActions(actions);
+        }
+        final List<JpaAction> savedActions = activateActions(actions);
+        setAssignmentOnTargets(savedActions);
+        return new ArrayList<>(savedActions);
+    }
 
-        if (isMultiAssignmentsEnabled()) {
-            overrideObsoleteUpdateActions = Collections.emptyList();
+    private void closeOrCancelOpenDeviceActions(final List<JpaAction> actions){
+        final List<Long> targetIds = actions.stream().map(JpaAction::getTarget).map(Target::getId)
+              .collect(Collectors.toList());
+        if (isActionsAutocloseEnabled()) {
+            onlineDsAssignmentStrategy.closeObsoleteUpdateActions(targetIds);
         } else {
-            final List<Long> targetId = Collections.singletonList(action.getTarget().getId());
-            if (isActionsAutocloseEnabled()) {
-                overrideObsoleteUpdateActions = Collections.emptyList();
-                onlineDsAssignmentStrategy.closeObsoleteUpdateActions(targetId);
-            } else {
-                overrideObsoleteUpdateActions = onlineDsAssignmentStrategy.overrideObsoleteUpdateActions(targetId);
-            }
+            onlineDsAssignmentStrategy.overrideObsoleteUpdateActions(targetIds);
         }
+    }
 
-        action.setActive(true);
-        action.setStatus(Status.RUNNING);
-        final JpaAction savedAction = actionRepository.save(action);
-
-        actionStatusRepository.save(onlineDsAssignmentStrategy.createActionStatus(savedAction, null));
-
-        final JpaTarget target = (JpaTarget) entityManager.merge(savedAction.getTarget());
-
-        target.setAssignedDistributionSet(savedAction.getDistributionSet());
-        target.setUpdateStatus(TargetUpdateStatus.PENDING);
-        targetRepository.save(target);
+    private List<JpaAction> activateActions(final List<JpaAction> actions){
+        actions.forEach(action -> {
+            action.setActive(true);
+            action.setStatus(Status.RUNNING);
+        });
+        return actionRepository.saveAll(actions);
+    }
 
-        // in case we canceled an action before for this target, then don't fire
-        // assignment event
-        if (overrideObsoleteUpdateActions.contains(savedAction.getId())) {
-            return null;
-        }
+    private void setAssignmentOnTargets(final List<JpaAction> actions){
+        final List<JpaTarget> assignedDsTargets = actions.stream().map(savedAction -> {
+            actionStatusRepository.save(onlineDsAssignmentStrategy.createActionStatus(savedAction, null));
```

Comment:
please check if it is feasible to bulk save all action statuses at once

Reply:
moved the persistence of the action status out of the stream to save all at once